pkg/fleet-manager/pipeline/controller.go

```diff
@@ -0,0 +1,355 @@
+			log.Info("Pipeline object not found", "pipeline", req)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Initialize a helper for patching the pipeline object at the end.
+	patchHelper, err := patch.NewHelper(pipeline, p.Client)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to init patch helper for pipeline %s", req.NamespacedName)
+	}
+	defer func() {
+		patchOpts := []patch.Option{}
+		if err := patchHelper.Patch(ctx, pipeline, patchOpts...); err != nil {
+			log.Error(err, "error patching pipeline")
+			reterr = utilerrors.NewAggregate([]error{reterr, errors.Wrapf(err, "failed to patch pipeline %s", req.NamespacedName)})
+		}
+	}()
+
+	// Add finalizer if it's not present.
+	if !controllerutil.ContainsFinalizer(pipeline, PipelineFinalizer) {
+		controllerutil.AddFinalizer(pipeline, PipelineFinalizer)
+		return ctrl.Result{}, nil
+	}
+
+	// Handle pipeline deletion.
+	if pipeline.GetDeletionTimestamp() != nil {
+		return p.reconcileDeletePipeline(ctx, pipeline)
+	}
+
+	// Proceed with the main reconciliation logic.
+	return p.reconcilePipeline(ctx, pipeline)
+}
+
+// reconcilePipeline contains the core logic for reconciling the Pipeline object.
+func (p *PipelineManager) reconcilePipeline(ctx context.Context, pipeline *pipelineapi.Pipeline) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Prepare RBAC configuration for the pipeline.
+	rbacConfig := render.RBACConfig{
+		PipelineName:         pipeline.Name,
+		PipelineNamespace:    pipeline.Namespace,
+		ChainCredentialsName: ChainCredentials,
+		OwnerReference:       render.GeneratePipelineOwnerRef(pipeline),
+	}
+
+	// Ensure RBAC resources are created before other resources.
+	if !p.isRBACResourceReady(ctx, rbacConfig) {
+		result, err := p.reconcileCreateRBAC(ctx, rbacConfig)
+		if err != nil || result.Requeue || result.RequeueAfter > 0 {
+			return result, err
+		}
+		// Add interval for creating rbac resource
+		return ctrl.Result{RequeueAfter: 1 * time.Second}, nil
+	}
+
+	// Create and apply Tekton tasks.
+	res, err := p.reconcileCreateTasks(ctx, pipeline)
+	if err != nil || res.Requeue || res.RequeueAfter > 0 {
+		log.Error(err, "Error creating Tekton tasks")
+		return res, err
+	}
+
+	// Create and apply Tekton pipeline.
+	res, err = p.reconcileCreatePipeline(ctx, pipeline)
+	if err != nil || res.Requeue || res.RequeueAfter > 0 {
+		return res, err
+	}
+
+	// Create and apply Tekton trigger.
+	res, err = p.reconcileCreateTrigger(ctx, pipeline)
+	if err != nil || res.Requeue || res.RequeueAfter > 0 {
+		return res, err
+	}
+
+	// Update pipeline status.
+	return p.reconcilePipelineStatus(ctx, pipeline)
+}
+
+// reconcileCreateRBAC creates and applies RBAC resources for the pipeline.
+func (p *PipelineManager) reconcileCreateRBAC(ctx context.Context, rbacConfig render.RBACConfig) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Render RBAC configuration.
+	rbac, err := render.RenderRBAC(rbacConfig)
```

Comment:
When it cane be changed, how could you update it?

Reply:
if we need update `rbac`, we should consider refactor `isRBACResourceReady` func

Reply:
we still need consider that the `rbac` should be updated in the front of others.

Reply:
I remove isRBACResourceReady func and everything is ok when I test this. It may be handled by Tekton.

then, now we can update it.